### PR TITLE
fix[lang]: typecheck hashmap indexes with folding

### LINKED ISSUE #3984: HashMap index checks when the subscript is folded
per @Leminkay

### Version Information

vyper Version (output of vyper --version): 6f09e2966b8096d8d36a85ad8745d56691dd25c4

### Issue description and/or POC
```vyper
m: HashMap[uint256, uint256]

@external
def foo():
   self.m[0-1] = 2
```
Successfully compiles

introduced in https://github.com/vyperlang/vyper/pull/3669, _related_ bug was fixed in https://github.com/vyperlang/vyper/pull/3924 (but this bug was not fixed)

### PR DESCRIPTION
**Pull request by DanielSchiavini**
Fixes #3984

### What I did
- Updated index checks for when the subscript is folded

### How I did it
- Use the reduced (i.e. folded) value when checking the index type

### How to verify it
- Tests are included

### Commit message
```
this commit fixes a bug in typechecking of folded values for hashmap
indexes. there was previously a carveout for subscript typechecking for
arrays; however, it should not apply to hashmap indexes.

this was introduced as an edge case resulting from the semantic changes
in 56c4c9dbc0. a related fix was applied in 75fb0594ab3, but it did not
handle hashmaps.
```

### Description for the changelog
`HashMap index is checked when the subscript is folded`

### Cute Animal Picture
image

### CODE REVIEW

File: vyper/semantics/analysis/local.py [hunk 1]
```diff
@@ -914,14 +914,14 @@ def visit_Subscript(self, node: vy_ast.Subscript, typ: VyperType) -> None:
...
 
-        # get the correct type for the index, it might
-        # not be exactly base_type.key_type
-        # note: index_type is validated in types_from_Subscript
-        index_types = get_possible_types_from_node(node.slice)
-        index_type = index_types.pop()
+        if isinstance(base_type, HashMapT):
+            index_type = base_type.key_type
+        else:
+            # Arrays allow most int types as index: Take the least specific
```
Comment (charles-cooper) [inline vyper/semantics/analysis/local.py:920]: maybe it's worth refactoring into a function. `base_type.index_type_from_node(node.slice)`

Comment (DanielSchiavini) [reply vyper/semantics/analysis/local.py:920]: that generates more import cycles 
image

**Review by charles-cooper — COMMENTED** [review]
this looks good to me. paging @Leminkay to double check

**Comment by Leminkay** [comment]
LGTM!

**Comment by charles-cooper** [comment]
@DanielSchiavini update the commit message please with a description of how the change was arrived at, and i'll merge

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**